spring-data-r2dbc/src/main/java/org/springframework/data/r2dbc/core/mapping/IdGeneratingBeforeSaveCallback.java

```diff
@@ -0,0 +1,103 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.data.r2dbc.core.mapping;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.jetbrains.annotations.NotNull;
+import org.reactivestreams.Publisher;
+import org.springframework.data.r2dbc.dialect.R2dbcDialect;
+import org.springframework.data.r2dbc.mapping.OutboundRow;
+import org.springframework.data.r2dbc.mapping.event.BeforeSaveCallback;
+import org.springframework.data.relational.core.mapping.RelationalMappingContext;
+import org.springframework.data.relational.core.mapping.RelationalPersistentEntity;
+import org.springframework.data.relational.core.sql.SqlIdentifier;
+import org.springframework.r2dbc.core.DatabaseClient;
+import org.springframework.r2dbc.core.Parameter;
+import org.springframework.util.Assert;
+
+import reactor.core.publisher.Mono;
+
+/**
+ * R2DBC Callback for generating ID via the database sequence.
+ *
+ * @author Mikhail Polivakha
+ */
+public class IdGeneratingBeforeSaveCallback implements BeforeSaveCallback<Object> {
+
+	private static final Log LOG = LogFactory.getLog(IdGeneratingBeforeSaveCallback.class);
+
+	private final RelationalMappingContext relationalMappingContext;
+	private final R2dbcDialect dialect;
+
+	private final DatabaseClient databaseClient;
+
+	public IdGeneratingBeforeSaveCallback(RelationalMappingContext relationalMappingContext, R2dbcDialect dialect,
+			DatabaseClient databaseClient) {
+		this.relationalMappingContext = relationalMappingContext;
+		this.dialect = dialect;
+		this.databaseClient = databaseClient;
+	}
+
+	@Override
+	public Publisher<Object> onBeforeSave(Object entity, OutboundRow row, SqlIdentifier table) {
+		Assert.notNull(entity, "The aggregate cannot be null at this point");
+
+		RelationalPersistentEntity<?> persistentEntity = relationalMappingContext.getPersistentEntity(entity.getClass());
+
+		if (!persistentEntity.hasIdProperty() || //
+				!persistentEntity.getIdProperty().hasSequence() || //
+				!persistentEntity.isNew(entity) //
+		) {
+			return Mono.just(entity);
+		}
+
+		SqlIdentifier idSequence = persistentEntity.getIdProperty().getSequence();
+
+		if (dialect.getIdGeneration().sequencesSupported()) {
+			return fetchIdFromSeq(entity, row, persistentEntity, idSequence);
+		} else {
+			illegalSequenceUsageWarning(entity);
+		}
+
+		return Mono.just(entity);
+	}
+
+	@NotNull
+	private Mono<Object> fetchIdFromSeq(Object entity, OutboundRow row, RelationalPersistentEntity<?> persistentEntity,
+			SqlIdentifier idSequence) {
+		String sequenceQuery = dialect.getIdGeneration().createSequenceQuery(idSequence);
+
+		return databaseClient //
+				.sql(sequenceQuery) //
+				.map((r, rowMetadata) -> r.get(0)) //
+				.one() //
+				.doOnNext(fetchedId -> { //
+					row.put(persistentEntity.getIdColumn().toSql(dialect.getIdentifierProcessing()), //
```

Comment:
Obtaining the IdColumn as early as possible guards the later code against potential absence. Also, there's repetative access to `IdProperty`. I think it would be much more aligned with a future support of embedded Id's to use the IdProperty instead of IdColumn.

Reply:
> I think it would be much more aligned with a future support of embedded Id's to use the IdProperty instead of IdColumn.

Can you please elaborate on what you mean here? [The `OutboundRow` contains the mapping between column names (not properties) and their repsective values](https://github.com/spring-projects/spring-data-relational/blob/main/spring-data-r2dbc/src/main/java/org/springframework/data/r2dbc/core/DefaultReactiveDataAccessStrategy.java#L188). I can obtain the `idProperty` once (although, this operation is relatively cheap since it just returns a reference to an already computed Id `PersistentProperty`), but I do not see how we can use the property here intead of column.

Reply:
This is a difference between `persistentEntity.getIdColumn()` and `PersistentProperty.getColumnName()`.

Reply:
I still do not understand what change needs to be done here to be honest @mp911de  

> This is a difference between persistentEntity.getIdColumn() and PersistentProperty.getColumnName().

The `persistentEntity.getIdColumn()` is implemeted internally as `getRequiredIdProperty().getColumnName()`, which partically the same as calling the `PersistentProperty.getColumnName()`.